File: lib/mydia_web/live/helpers/grid_density.ex
defmodule MydiaWeb.Live.Helpers.GridDensity do
  @moduledoc """
  Reads and sets the poster-grid density for the Discover and Libraries
  LiveViews.

  Density is a per-browser setting, not an account preference, so a
  desktop's choice never reaches the same account's phone (#700). The
  browser holds it in the `mydia_grid_density` cookie:

    * `MydiaWeb.Plugs.GridDensityCookie` copies the cookie into the session
      on every HTTP request, and `assign_current/2` reads it from there.
    * `put/2` pushes `grid_density:saved`, and the `GridDensity` JS hook
      writes the cookie only then, so the browser stores only values the
      server accepted.

  The session is fixed when the LiveView socket connects, so after a change
  a live navigation to the other page mounts with the old value. The hook
  compares the cookie with the rendered toggle on mount and re-sends
  `set_grid_density` when they differ.
  """

  import Phoenix.Component, only: [assign: 3]
  import Phoenix.LiveView, only: [push_event: 3]

  alias MydiaWeb.GridDensityComponents

  @assign :grid_density

  @doc """
  Assigns this browser's density from the session, or the default.
  """
  def assign_current(socket, session) do
    assign(socket, @assign, current(session))
  end

  @doc """
  This browser's density from the session, or the default.
  """
  def current(%{"grid_density" => density}) do
    if GridDensityComponents.valid?(density),
      do: density,
      else: GridDensityComponents.default()
  end

  def current(_session), do: GridDensityComponents.default()

  @doc """
  Applies a density to the current view and asks the browser to keep it.
  An unknown value is ignored.
  """
  def put(socket, density) do
    if GridDensityComponents.valid?(density) do
      socket
      |> assign(@assign, density)
      |> push_event("grid_density:saved", %{density: density})
    else
      socket
    end
  end
end

File: assets/js/hooks/grid_density.mjs
// Keeps the poster-grid density in a per-browser cookie.
//
// The server decides: a toggle click goes to the LiveView, and only once it
// accepts the value does it push `grid_density:saved`, which is when the
// cookie is written. MydiaWeb.Plugs.GridDensityCookie reads the cookie on
// the next HTTP request so the first render already has the right columns.
//
// The LiveView session is fixed when the socket connects. After changing
// density on one page, a live navigation to the other mounts with the old
// value, so on mount the hook compares the cookie with the rendered toggle
// and re-sends the stored value when they differ.

export const COOKIE_NAME = "mydia_grid_density";
export const LEVELS = ["comfortable", "compact", "dense"];

const ONE_YEAR_SECONDS = 60 * 60 * 24 * 365;

export function readDensityCookie(cookieString) {
  for (const part of (cookieString || "").split(";")) {
    const [name, ...rest] = part.trim().split("=");

    if (name === COOKIE_NAME) {
      const value = rest.join("=");
      return LEVELS.includes(value) ? value : null;
    }
  }

  return null;
}

export function densityCookie(value) {
  return `${COOKIE_NAME}=${value}; path=/; max-age=${ONE_YEAR_SECONDS}; SameSite=Lax`;
}

const GridDensity = {
  mounted() {
    this.handleEvent("grid_density:saved", ({ density }) => {
      if (LEVELS.includes(density)) document.cookie = densityCookie(density);
    });

    const stored = readDensityCookie(document.cookie);

    if (stored && stored !== this.el.dataset.value) {
      this.pushEvent("set_grid_density", { density: stored });
    }
  },
};

export default GridDensity;
